src/manifold/src/smoothing.cpp

```diff
@@ -308,6 +191,296 @@ struct InterpTri {
+                    {edgeOffsets[1] - 1, edgeOffsets[1], 2, 0},
+                    {-1, edgeOffsets[1] + 1, edgeOffsets[2], edgeOffsets[0]},
+                    {0, n[1] - 2, n[2] - 1, n[0] - 2},
+                    {true, true, true, true});
+    } else {  // obtuse -> spit into two acute
+      // portion of n[0] under n[2]
+      const int ns = (f - n[1] * n[1]) / (2 * n[0]);
+      // height from n[0]: nh <= n[2]
+      const int nh = glm::max(1., glm::round(glm::sqrt(n[2] * n[2] - ns * ns)));
+
+      const int hOffset = partition.vertBary.size();
+      const glm::vec3 middleBary = partition.vertBary[edgeOffsets[0] + ns - 1];
+      for (int j = 1; j < nh; ++j) {
+        partition.vertBary.push_back(
+            glm::mix(partition.vertBary[2], middleBary, (float)j / nh));
+      }
+
+      partition.triVert.push_back({edgeOffsets[1] - 1, 1, edgeOffsets[1]});
+      PartitionQuad(
+          partition.triVert, partition.vertBary,
+          {edgeOffsets[1] - 1, edgeOffsets[1], 2, edgeOffsets[0] + ns - 1},
+          {-1, edgeOffsets[1] + 1, hOffset, edgeOffsets[0] + ns},
+          {0, n[1] - 2, nh - 1, n[0] - ns - 2}, {true, true, true, true});
+
+      if (n[2] == 1) {
+        PartitionFan(partition.triVert, {0, edgeOffsets[0] + ns - 1, 2}, ns - 1,
+                     edgeOffsets[0]);
+      } else {
+        if (ns == 1) {
+          partition.triVert.push_back({hOffset, 2, edgeOffsets[2]});
+          PartitionQuad(partition.triVert, partition.vertBary,
+                        {hOffset, edgeOffsets[2], 0, edgeOffsets[0]},
+                        {-1, edgeOffsets[2] + 1, -1, hOffset + nh - 2},
+                        {0, n[2] - 2, ns - 1, nh - 2},
+                        {true, true, true, false});
+        } else {
+          partition.triVert.push_back({hOffset - 1, 0, edgeOffsets[0]});
+          PartitionQuad(
+              partition.triVert, partition.vertBary,
+              {hOffset - 1, edgeOffsets[0], edgeOffsets[0] + ns - 1, 2},
+              {-1, edgeOffsets[0] + 1, hOffset + nh - 2, edgeOffsets[2]},
+              {0, ns - 2, nh - 1, n[2] - 2}, {true, true, false, true});
+        }
+      }
+    }
+
+    cache.insert({n, partition});
+    return partition;
+  }
+
+  // Side 0 has added edges while sides 1 and 2 do not. Fan spreads from vert 2.
+  static void PartitionFan(Vec<glm::ivec3>& triVert, glm::ivec3 cornerVerts,
+                           int added, int edgeOffset) {
+    int last = cornerVerts[0];
+    for (int i = 0; i < added; ++i) {
+      const int next = edgeOffset + i;
+      triVert.push_back({last, next, cornerVerts[2]});
+      last = next;
+    }
+    triVert.push_back({last, cornerVerts[1], cornerVerts[2]});
+  }
+
+  // Partitions are parallel to the first edge unless two consecutive edgeAdded
+  // are zero, in which case a terminal triangulation is performed.
+  static void PartitionQuad(Vec<glm::ivec3>& triVert, Vec<glm::vec3>& vertBary,
```

Comment:
The algorithm actually focuses on partitioning quads recursively - triangles are turned into quads by clipping one sub-triangle off of a corner.